# Pembersihan dan Prapemrosesan data

In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn_pandas import CategoricalImputer

In [4]:
df = pd.read_csv("top50contry.csv", encoding="ISO-8859-1")

df

Unnamed: 0                            title          artist  \
0             1                     Dance Monkey     Tones and I   
1             2                          ROXANNE  Arizona Zervas   
2             3                         Memories        Maroon 5   
3             4                          Circles     Post Malone   
4             5  All I Want for Christmas Is You    Mariah Carey   
..          ...                              ...             ...   
995         996                         Feelings            Lauv   
996         997               Mimpi (feat. Alif)        K-Clique   
997         998                 Cinta Luar Biasa         Andmesh   
998         999                        Haa Tepok         MeerFly   
999        1000                      Hanya Rindu         Andmesh   

             top genre  year       added    bpm  nrgy  dnce    dB  live   val  \
0       australian pop  2019  1969-12-31   98.0  59.0  82.0  -6.0  15.0  51.0   
1              pop rap  2019  1969-12-31  117.0  60.0  62.0  -6.0  46.0  46.0   
2                  pop  2019  1969-12-31   91.0  32.0  76.0  -7.0   8.0  57.0   
3              dfw rap  2019  1969-12-31  120.0  76.0  70.0  -3.0   9.0  55.0   
4            dance pop  1994  1969-12-31  150.0  63.0  34.0  -7.0   7.0  35.0   
..                 ...   ...         ...    ...   ...   ...   ...   ...   ...   
995                pop  2019  1969-12-31  103.0  47.0  69.0  -7.0  10.0  25.0   
996    malaysian indie  2019  1969-12-31  150.0  73.0  83.0  -7.0  10.0  87.0   
997     indonesian pop  2019  1969-12-31  136.0  23.0  58.0 -12.0  10.0  32.0   
998  malaysian hip hop  2019  1969-12-31   87.0  50.0  85.0 -12.0  15.0  68.0   
999     indonesian pop  2019  1969-12-31  179.0  16.0  28.0 -12.0  11.0  20.0   

     dur  acous  spch  pop  country  
0    209   69.0   9.0  100    world  
1    164    5.0  15.0   99    world  
2    189   84.0   5.0   99    world  
3    215   19.0   4.0   99    world  
4    241   16.0   4.0   98    world  
..   ...    ...   ...  ...      ...  
995  190    9.0   4.0   73  malasya  
996  253   53.0  29.0   66  malasya  
997  256   81.0   3.0   66  malasya  
998  211   66.0  31.0   65  malasya  
999  252   89.0   3.0   65  malasya  

[1000 rows x 17 columns]

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 17 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  1000 non-null   int64  
 1   title       1000 non-null   object 
 2   artist      1000 non-null   object 
 3   top genre   985 non-null    object 
 4   year        1000 non-null   int64  
 5   added       1000 non-null   object 
 6   bpm         999 non-null    float64
 7   nrgy        999 non-null    float64
 8   dnce        999 non-null    float64
 9   dB          999 non-null    float64
 10  live        999 non-null    float64
 11  val         999 non-null    float64
 12  dur         1000 non-null   int64  
 13  acous       999 non-null    float64
 14  spch        999 non-null    float64
 15  pop         1000 non-null   int64  
 16  country     1000 non-null   object 
dtypes: float64(8), int64(4), object(5)
memory usage: 132.9+ KB


In [6]:
df.describe()

Unnamed: 0         year         bpm        nrgy        dnce  \
count  1000.000000  1000.000000  999.000000  999.000000  999.000000   
mean    500.500000  2010.025000  123.644645   61.650651   64.282282   
std     288.819436    16.904264   31.005120   19.064732   16.032857   
min       1.000000  1942.000000   47.000000   10.000000   16.000000   
25%     250.750000  2011.000000   98.000000   49.000000   53.000000   
50%     500.500000  2019.000000  120.000000   65.000000   67.000000   
75%     750.250000  2019.000000  147.000000   76.000000   76.000000   
max    1000.000000  2019.000000  205.000000   98.000000   95.000000   

               dB        live         val          dur       acous  \
count  999.000000  999.000000  999.000000  1000.000000  999.000000   
mean    -6.807808   20.003003   59.732733   206.074000   34.894895   
std      3.241313   17.956443   22.000356    51.224198   28.778784   
min    -23.000000    2.000000    5.000000    85.000000    0.000000   
25%     -8.000000    9.000000   42.000000   172.000000   10.000000   
50%     -6.000000   12.000000   61.000000   201.000000   27.000000   
75%     -5.000000   25.000000   77.000000   231.000000   59.000000   
max      0.000000   98.000000   98.000000   464.000000   99.000000   

             spch          pop  
count  999.000000  1000.000000  
mean     8.703704    81.443000  
std      8.432230    13.090229  
min      2.000000     0.000000  
25%      4.000000    75.000000  
50%      5.000000    85.000000  
75%      9.000000    90.000000  
max     56.000000   100.000000

## NORMALISASI NAMA KOLOM

Dari dataframe yang ditampilkan di atas, kita memiliki satu kolom yang paling awal (paling kiri) yang tidak bernama (unnamed), kolom yang tidak bernama ini akan menyulitkan proses analisis data. Maka, ada dua alternatif solusi:

1. Menghapus (drop) kolom tersebut
2. Memberikan nama pada kolom tersebut

Dalam dataset, terlihat bahwa konteks untuk kolom pertama tersebut adalah ID dari lagunya. Maka, untuk sementara, kolom tersebut bisa diberi nama "ID"

In [7]:
df.columns.tolist()

['Unnamed: 0',
 'title',
 'artist',
 'top genre',
 'year',
 'added',
 'bpm',
 'nrgy',
 'dnce',
 'dB',
 'live',
 'val',
 'dur',
 'acous',
 'spch',
 'pop',
 'country']

In [8]:
df.rename(columns = {"Unnamed: 0" : "ID"}, inplace=True)
df

ID                            title          artist          top genre  \
0       1                     Dance Monkey     Tones and I     australian pop   
1       2                          ROXANNE  Arizona Zervas            pop rap   
2       3                         Memories        Maroon 5                pop   
3       4                          Circles     Post Malone            dfw rap   
4       5  All I Want for Christmas Is You    Mariah Carey          dance pop   
..    ...                              ...             ...                ...   
995   996                         Feelings            Lauv                pop   
996   997               Mimpi (feat. Alif)        K-Clique    malaysian indie   
997   998                 Cinta Luar Biasa         Andmesh     indonesian pop   
998   999                        Haa Tepok         MeerFly  malaysian hip hop   
999  1000                      Hanya Rindu         Andmesh     indonesian pop   

     year       added    bpm  nrgy  dnce    dB  live   val  dur  acous  spch  \
0    2019  1969-12-31   98.0  59.0  82.0  -6.0  15.0  51.0  209   69.0   9.0   
1    2019  1969-12-31  117.0  60.0  62.0  -6.0  46.0  46.0  164    5.0  15.0   
2    2019  1969-12-31   91.0  32.0  76.0  -7.0   8.0  57.0  189   84.0   5.0   
3    2019  1969-12-31  120.0  76.0  70.0  -3.0   9.0  55.0  215   19.0   4.0   
4    1994  1969-12-31  150.0  63.0  34.0  -7.0   7.0  35.0  241   16.0   4.0   
..    ...         ...    ...   ...   ...   ...   ...   ...  ...    ...   ...   
995  2019  1969-12-31  103.0  47.0  69.0  -7.0  10.0  25.0  190    9.0   4.0   
996  2019  1969-12-31  150.0  73.0  83.0  -7.0  10.0  87.0  253   53.0  29.0   
997  2019  1969-12-31  136.0  23.0  58.0 -12.0  10.0  32.0  256   81.0   3.0   
998  2019  1969-12-31   87.0  50.0  85.0 -12.0  15.0  68.0  211   66.0  31.0   
999  2019  1969-12-31  179.0  16.0  28.0 -12.0  11.0  20.0  252   89.0   3.0   

     pop  country  
0    100    world  
1     99    world  
2     99    world  
3     99    world  
4     98    world  
..   ...      ...  
995   73  malasya  
996   66  malasya  
997   66  malasya  
998   65  malasya  
999   65  malasya  

[1000 rows x 17 columns]

Untuk memastikan bahwa konteks kolom pertama tersebut benar adalah ID (unik untuk setiap lagu), maka kita bisa periksa apakah judul lagu yang sama memiliki ID yang sama. Sebagai contoh (dan agar penjelasan lebih intuitif) akan dianalisa ID dari lagu "Dance Monkey"

In [9]:
df.loc[df["title"]=="Dance Monkey"]

ID         title       artist       top genre  year       added   bpm  \
0      1  Dance Monkey  Tones and I  australian pop  2019  1969-12-31  98.0   
50    51  Dance Monkey  Tones and I  australian pop  2019  1969-12-31  98.0   
100  101  Dance Monkey  Tones and I  australian pop  2019  1969-12-31  98.0   
150  151  Dance Monkey  Tones and I  australian pop  2019  1969-12-31  98.0   
200  201  Dance Monkey  Tones and I  australian pop  2019  1969-12-31  98.0   
250  251  Dance Monkey  Tones and I  australian pop  2019  1969-12-31  98.0   
300  301  Dance Monkey  Tones and I  australian pop  2019  1969-12-31  98.0   
350  351  Dance Monkey  Tones and I  australian pop  2019  1969-12-31  98.0   
400  401  Dance Monkey  Tones and I  australian pop  2019  1969-12-31  98.0   
450  451  Dance Monkey  Tones and I  australian pop  2019  1969-12-31  98.0   
500  501  Dance Monkey  Tones and I  australian pop  2019  1969-12-31  98.0   
550  551  Dance Monkey  Tones and I  australian pop  2019  1969-12-31  98.0   
650  651  Dance Monkey  Tones and I  australian pop  2019  1969-12-31  98.0   
750  751  Dance Monkey  Tones and I  australian pop  2019  1969-12-31  98.0   
800  801  Dance Monkey  Tones and I  australian pop  2019  1969-12-31  98.0   
850  851  Dance Monkey  Tones and I  australian pop  2019  1969-12-31  98.0   
950  951  Dance Monkey  Tones and I  australian pop  2019  1969-12-31  98.0   

     nrgy  dnce   dB  live   val  dur  acous  spch  pop    country  
0    59.0  82.0 -6.0  15.0  51.0  209   69.0   9.0  100      world  
50   59.0  82.0 -6.0  15.0  51.0  209   69.0   9.0  100     africa  
100  59.0  82.0 -6.0  15.0  51.0  209   69.0   9.0  100  argentina  
150  59.0  82.0 -6.0  15.0  51.0  209   69.0   9.0  100  australia  
200  59.0  82.0 -6.0  15.0  51.0  209   69.0   9.0  100    belgium  
250  59.0  82.0 -6.0  15.0  51.0  209   69.0   9.0  100    bolivia  
300  59.0  82.0 -6.0  15.0  51.0  209   69.0   9.0  100     brazil  
350  59.0  82.0 -6.0  15.0  51.0  209   69.0   9.0  100     canada  
400  59.0  82.0 -6.0  15.0  51.0  209   69.0   9.0  100    germany  
450  59.0  82.0 -6.0  15.0  51.0  209   69.0   9.0  100   colombia  
500  59.0  82.0 -6.0  15.0  51.0  209   69.0   9.0  100      chile  
550  59.0  82.0 -6.0  15.0  51.0  209   69.0   9.0  100      spain  
650  59.0  82.0 -6.0  15.0  51.0  209   69.0   9.0  100     france  
750  59.0  82.0 -6.0  15.0  51.0  209   69.0   9.0  100  indonesia  
800  59.0  82.0 -6.0  15.0  51.0  209   69.0   9.0  100     israel  
850  59.0  82.0 -6.0  15.0  51.0  209   69.0   9.0  100      italy  
950  59.0  82.0 -6.0  15.0  51.0  209   69.0   9.0  100    malasya

Dari hasil di atas terlihat bahwa lagu yang sama, yaitu "Dance Monkey" ciptaan Tones and I, memiliki ID yang berbeda-beda, padahal secara intuitif ID lagu seharusnya unik untuk setiap lagu yang unik (berbeda), bukan untuk setiap entri yang berbeda. Oleh karena itu, dirasa kolom ID ini kurang cocok untuk konteks ID, lebih baik dinamakan kolom "nomor". Tetapi, mengingat bahwa dataframe sudah memiliki indeks, sebenarnya kolom "nomor" tidak diperlukan (redundant).

Dari pemaparan di atas, bisa disimpulkan bahwa kolom ID bisa dihapus karena tidak diperlukan.

In [10]:
df = df.drop("ID", axis=1)

## PEMERIKSAAN DATA DUPLIKAT

Selanjutnya, kita mencari entri data yang benar-benar duplikat, benar-benar duplikat dalam konteks dataset ini adalah bahwa **SELURUH** atribut pada satu lagu atau entri benar-benar sama **PERSIS**. Mengapa demikian? Karena bila duplikat didefinisikan sebagai entri atau lagu yang memiliki **SALAH SATU** atribut yang sama, maka lagu atau entri yang memiliki judul yang sama tetapi popular di negara berbeda akan terhitung sebagai duplikat, padahal mereka mendeskripsikan ketenaran lagu pada negara yang berbeda.

Maka untuk memeriksa kegandaan atau duplikat yang **SELURUH** atributnya sama persis, bisa dilakukan dengan proses di bawah

In [11]:
df.duplicated()

0      False
1      False
2      False
3      False
4      False
       ...  
995    False
996    False
997    False
998    False
999    False
Length: 1000, dtype: bool

In [12]:
df[df.duplicated()]

Empty DataFrame
Columns: [title, artist, top genre, year, added, bpm, nrgy, dnce, dB, live, val, dur, acous, spch, pop, country]
Index: []

Terlihat dari program di atas bahwa setiap entri adalah **UNIK**, tidak ada entri yang benar-benar duplikat. Sehingga, bisa dilanjutkan ke tahap cleaning selanjutnya, yaitu pemeriksaan nilai null atau NA.

## PEMERIKSAAN NILAI NULL

In [13]:
df.loc[df.isnull().any(axis=1)]

title             artist  \
29         Do They Know It's Christmas? - 1984 Version           Band Aid   
72         Do They Know It's Christmas? - 1984 Version           Band Aid   
96                                                Fire         DJ Big Sky   
136                                          Ya No Mas        Fer Palacio   
170        Do They Know It's Christmas? - 1984 Version           Band Aid   
228        Do They Know It's Christmas? - 1984 Version           Band Aid   
290                                               Loco              Beéle   
370        Do They Know It's Christmas? - 1984 Version           Band Aid   
416        Do They Know It's Christmas? - 1984 Version           Band Aid   
433  Wonderful Dream (Holidays Are Coming) - Radio ...   Melanie Thornton   
477                                               Loco              Beéle   
623        Do They Know It's Christmas? - 1984 Version           Band Aid   
749                  Lamborghini (From "Jai Mummy Di")         Meet Bros.   
770                                               Halu         Feby Putri   
790                             Kartonyono Medot Janji       Denny Caknan   
791                                        Percaya Aku  Chintya Gabriella   

    top genre  year       added    bpm  nrgy  dnce    dB  live   val  dur  \
29        NaN  2004  1969-12-31  115.0  64.0  60.0  -7.0  10.0  35.0  223   
72        NaN  2004  1969-12-31  115.0  64.0  60.0  -7.0  10.0  35.0  223   
96        NaN  2019  1969-12-31  114.0  56.0  65.0  -7.0   5.0  64.0  463   
136       NaN  2019  1969-12-31  100.0  68.0  80.0  -6.0   7.0  88.0  158   
170       NaN  2004  1969-12-31  115.0  64.0  60.0  -7.0  10.0  35.0  223   
228       NaN  2004  1969-12-31  115.0  64.0  60.0  -7.0  10.0  35.0  223   
290       NaN  2019  1969-12-31  105.0  56.0  82.0  -6.0   5.0  77.0  204   
370       NaN  2004  1969-12-31  115.0  64.0  60.0  -7.0  10.0  35.0  223   
416       NaN  2004  1969-12-31  115.0  64.0  60.0  -7.0  10.0  35.0  223   
433       NaN  2001  1969-12-31  150.0  68.0  58.0  -5.0  26.0  52.0  232   
477       NaN  2019  1969-12-31  105.0  56.0  82.0  -6.0   5.0  77.0  204   
623       NaN  2004  1969-12-31  115.0  64.0  60.0  -7.0  10.0  35.0  223   
749  desi pop  2019  1969-12-31    NaN   NaN   NaN   NaN   NaN   NaN  247   
770       NaN  2019  1969-12-31  116.0  25.0  34.0 -15.0  11.0  10.0  266   
790       NaN  2019  1969-12-31  130.0  78.0  73.0  -6.0  21.0  75.0  262   
791       NaN  2019  1969-12-31  140.0  49.0  54.0  -8.0  20.0  30.0  232   

     acous  spch  pop    country  
29    16.0   4.0   87      world  
72    16.0   4.0   87     africa  
96     2.0  21.0   47     africa  
136   14.0   6.0   75  argentina  
170   16.0   4.0   87  australia  
228   16.0   4.0   87    belgium  
290   16.0  17.0   82    bolivia  
370   16.0   4.0   87     canada  
416   16.0   4.0   87    germany  
433   11.0   3.0   79    germany  
477   16.0  17.0   82   colombia  
623   16.0   4.0   87        usa  
749    NaN   NaN   48      india  
770   87.0   4.0   79  indonesia  
790   31.0   3.0   71  indonesia  
791   74.0   3.0   71  indonesia

Melihat dari hasil di atas, kolom dengan nilai kosong paling banyak adalah kolom "top genre", dengan 15 nilai kosong dan entri dengan atribut kosong paling banyak adalah entri pada index 749, dengan 8 nilai kosong.

Ada 2 opsi untuk mengatasi nilai kosong:
1. Menghapus entri/kolom yang memiliki nilai kosong
2. Mengisi nilai kosong tersebut dengan suatu nilai lain (imputing)

Untuk nilai kosong pada entri dengan index 749 lebih masuk akal dilakukan penghapusan, karena data yang dimiliki sangat banyak (1000 entri) dan menghapus satu entri untuk negara India tidak akan terlalu memengaruhi analisis.

Sedangkan untuk nilai kosong pada kolom "top genre" lebih masuk akal dilakukan pengisian karena dataset hanya memiliki 2 kolom atribut kategorikal, sehingga penghapusan kolom "top genre" akan mempersulit analisis karena kekurangan atribut kategorikal.

Penanganan akan dimulai dengan nilai kosong pada entri index 749, tentukan terlebih dahulu urutan entri (lagu) tersebut pada urutan popularitas di negara India.

In [14]:
df_India = df.loc[df["country"]=="india"]
df_India.sort_values("pop", ascending=0)

title             artist  \
700                                       Memories           Maroon 5   
701                                        ROXANNE     Arizona Zervas   
702                                        Circles        Post Malone   
703                                        Falling      Trevor Daniel   
704                                Blinding Lights         The Weeknd   
705                                        bad guy      Billie Eilish   
706                                hot girl bummer          blackbear   
707                                      Heartless         The Weeknd   
708                         Trampoline (with ZAYN)              SHAED   
709                        My Oh My (feat. DaBaby)     Camila Cabello   
710                                       Señorita       Shawn Mendes   
711                                       Believer    Imagine Dragons   
712  Sunflower - Spider-Man: Into the Spider-Verse        Post Malone   
713                                      On My Way        Alan Walker   
714                          Ghungroo (From "War")       Arijit Singh   
715                Tum Hi Aana (From "Marjaavaan")          Payal Dev   
716                                        Filhall            B Praak   
717                                         Makhna     Tanishk Bagchi   
718                         Yaad Piya Ki Aane Lagi        Neha Kakkar   
719                   Duniyaa (From "Luka Chuppi")              Akhil   
720                    Pachtaoge (From "Jaani Ve")       Arijit Singh   
721   Tujhe Kitna Chahne Lage (From "Kabir Singh")       Arijit Singh   
722               Pal Pal Dil Ke Paas- Title Track       Arijit Singh   
723                                      Sakhiyaan    Maninder Buttar   
724                                  Proper Patola            Badshah   
729                                    Chal Bombay             DIVINE   
732                           Kalank (Title Track)       Arijit Singh   
731                                         Vaaste  Dhvani Bhanushali   
730                                        Ik Tera    Maninder Buttar   
725                 Bekhayali (From "Kabir Singh")      Sachet Tandon   
728               Jai Jai Shivshankar (From "War")     Vishal Dadlani   
726      Dheeme Dheeme (From "Pati Patni Aur Woh")        Tony Kakkar   
727                                       8 Parche       Baani Sandhu   
733          Shaitan Ka Saala (From "Housefull 4")         Sohail Sen   
734                      Naah Goriye (From "Bala")            B Praak   
735       Ek Toh Kum Zindagani (From "Marjaavaan")        Neha Kakkar   
736               Don't Be Shy Again (From "Bala")       Sachin-Jigar   
739                                Chandigarh Mein            Badshah   
740                                     Dance Like      Harrdy Sandhu   
738    Tu Hi Yaar Mera (From "Pati Patni Aur Woh")             Rochak   
737              Mere Sohneya (From "Kabir Singh")      Sachet Tandon   
741                              Sauda Khara Khara     Diljit Dosanjh   
742                 Kaise Hua (From "Kabir Singh")      Vishal Mishra   
743              Muqabla (From "Street Dancer 3D")      Yash Narvekar   
744                                       Ve Maahi       Arijit Singh   
745               O Saki Saki (From "Batla House")        Neha Kakkar   
746           Tera Ban Jaunga (From "Kabir Singh")     Akhil Sachdeva   
747          Mummy Nu Pasand (From "Jai Mummy Di")     Sunanda Sharma   
748            Hauli Hauli (From "De De Pyaar De")       Garry Sandhu   
749              Lamborghini (From "Jai Mummy Di")         Meet Bros.   

                     top genre  year       added    bpm  nrgy  dnce    dB  \
700                        pop  2019  1969-12-31   91.0  32.0  76.0  -7.0   
701                    pop rap  2019  1969-12-31  117.0  60.0  62.0  -6.0   
702                    dfw rap  2019  1969-12-31  120.0  76.0  70.0  -3.0   
703          

dilihat dari output program di atas, entri dengan atribut NaN yang banyak adalah lagu dengan ranking terendah dari negara India. Jadi tidak ada urutan di bawah atau di atasnya yang akan terpengaruh bila dilakukan penghapusan. Melihat kasus ini, maka opsi penghapusan masuk akal untuk data ini. Penghapusan akan dilakukan di akhir, bersama dengan penghapusan data yang irelevan dengan analisis.

selanjutnya, untuk kolom "top genre", *imputing* akan menggunakan categorical impute pada dari modul sklearn_pandas, seperti berikut

In [15]:
cat_imputer = CategoricalImputer()
df["top genre"] = pd.Series(cat_imputer.fit_transform(df["top genre"]))
df["top genre"].isnull().sum()

0

sekarang bisa kita cek apakah masih ada data NaN pada kolom "top genre"

In [16]:
df.loc[df.isnull().any(axis=1)]

title      artist top genre  year  \
749  Lamborghini (From "Jai Mummy Di")  Meet Bros.  desi pop  2019   

          added  bpm  nrgy  dnce  dB  live  val  dur  acous  spch  pop country  
749  1969-12-31  NaN   NaN   NaN NaN   NaN  NaN  247    NaN   NaN   48   india

Bisa dilihat, telah berhasil dilakukan imputing untuk menggantikan nilai-nilai NaN pada kolom "top genre"

## MENGHAPUS ENTRI DAN ATRIBUT YANG IRELEVAN

In [17]:
df

title          artist          top genre  year  \
0                       Dance Monkey     Tones and I     australian pop  2019   
1                            ROXANNE  Arizona Zervas            pop rap  2019   
2                           Memories        Maroon 5                pop  2019   
3                            Circles     Post Malone            dfw rap  2019   
4    All I Want for Christmas Is You    Mariah Carey          dance pop  1994   
..                               ...             ...                ...   ...   
995                         Feelings            Lauv                pop  2019   
996               Mimpi (feat. Alif)        K-Clique    malaysian indie  2019   
997                 Cinta Luar Biasa         Andmesh     indonesian pop  2019   
998                        Haa Tepok         MeerFly  malaysian hip hop  2019   
999                      Hanya Rindu         Andmesh     indonesian pop  2019   

          added    bpm  nrgy  dnce    dB  live   val  dur  acous  spch  pop  \
0    1969-12-31   98.0  59.0  82.0  -6.0  15.0  51.0  209   69.0   9.0  100   
1    1969-12-31  117.0  60.0  62.0  -6.0  46.0  46.0  164    5.0  15.0   99   
2    1969-12-31   91.0  32.0  76.0  -7.0   8.0  57.0  189   84.0   5.0   99   
3    1969-12-31  120.0  76.0  70.0  -3.0   9.0  55.0  215   19.0   4.0   99   
4    1969-12-31  150.0  63.0  34.0  -7.0   7.0  35.0  241   16.0   4.0   98   
..          ...    ...   ...   ...   ...   ...   ...  ...    ...   ...  ...   
995  1969-12-31  103.0  47.0  69.0  -7.0  10.0  25.0  190    9.0   4.0   73   
996  1969-12-31  150.0  73.0  83.0  -7.0  10.0  87.0  253   53.0  29.0   66   
997  1969-12-31  136.0  23.0  58.0 -12.0  10.0  32.0  256   81.0   3.0   66   
998  1969-12-31   87.0  50.0  85.0 -12.0  15.0  68.0  211   66.0  31.0   65   
999  1969-12-31  179.0  16.0  28.0 -12.0  11.0  20.0  252   89.0   3.0   65   

     country  
0      world  
1      world  
2      world  
3      world  
4      world  
..       ...  
995  malasya  
996  malasya  
997  malasya  
998  malasya  
999  malasya  

[1000 rows x 16 columns]

Data yang dimiliki sekarang terdiri dari 1000 entri dan 16 kolom, kita bisa analisa kolom-kolom dan entri-entri yang bisa dihapus berdasarkan:
1. Untuk entri, bisa dihapus bila dirasa data yang dimiliki terlalu banyak dan analisisnya sendiri tidak memerlukan data sebanyak itu
2. Untuk kolom kuantitatif, kolom dengan varians (variance) data yang rendah tidak menambahkan *insight* yang signifikan terhadap analisis.
3. Untuk kolom kategorikal, kolom dengan kategori entri paling seragam tidak memberikan *insight* yang signifikan terhadap analisis.

untuk entri, analis merasa bahwa top 50 lagu untuk setiap negara terlalu banyak, diputuskan untuk diambil setengahnya saja, sehingga akan difilter hanya top 25 lagu tiap negara.

selanjutnya untuk kolom, perhitungan variance akan digunakan method var() untuk pandas dataframe. Sedangkan keseragaman kategorinya akan dilihat dari value_counts()

### Kalkulasi varians

Pertama akan dihitung varians dari kolom kuantitatif, kolom "year" dikecualikan dalam kalkulasi karena analis akan memerlukan kolom tersebut sebagai salah satu kolom berisi *time series*

In [18]:
variance_array = []
df_kuanti = df.iloc[:, 5:14]
for column in df_kuanti.columns.tolist():
  variance_array.append(df_kuanti[column].var())
variance_array

[961.3174838164819,
 363.46400508725156,
 257.0525013991948,
 10.506111321742585,
 322.4338587084078,
 484.01567098160285,
 2623.9184424424425,
 828.2184007654951,
 71.1025012988941]

In [19]:
np.mean(variance_array)

658.0032195357237

untuk melihat kolom mana yang perlu dihapus, akan digunakan rata-rata variansnya sebagai patokan. Kolom dengan varians di bawah rata-rata akan dihapus

In [20]:
index_kolom_gagal = []
for variance in variance_array:
  if variance < np.mean(variance_array):
    index_kolom_gagal.append(variance_array.index(variance))

nama_kolom_gagal = []
for i in range(len(df_kuanti.columns.tolist())):
  if i in index_kolom_gagal:
    nama_kolom_gagal.append(df_kuanti.columns.tolist()[i])

nama_kolom_gagal

['nrgy', 'dnce', 'dB', 'live', 'val', 'spch']

In [21]:
df = df.drop(nama_kolom_gagal, axis=1)
df

title          artist          top genre  year  \
0                       Dance Monkey     Tones and I     australian pop  2019   
1                            ROXANNE  Arizona Zervas            pop rap  2019   
2                           Memories        Maroon 5                pop  2019   
3                            Circles     Post Malone            dfw rap  2019   
4    All I Want for Christmas Is You    Mariah Carey          dance pop  1994   
..                               ...             ...                ...   ...   
995                         Feelings            Lauv                pop  2019   
996               Mimpi (feat. Alif)        K-Clique    malaysian indie  2019   
997                 Cinta Luar Biasa         Andmesh     indonesian pop  2019   
998                        Haa Tepok         MeerFly  malaysian hip hop  2019   
999                      Hanya Rindu         Andmesh     indonesian pop  2019   

          added    bpm  dur  acous  pop  country  
0    1969-12-31   98.0  209   69.0  100    world  
1    1969-12-31  117.0  164    5.0   99    world  
2    1969-12-31   91.0  189   84.0   99    world  
3    1969-12-31  120.0  215   19.0   99    world  
4    1969-12-31  150.0  241   16.0   98    world  
..          ...    ...  ...    ...  ...      ...  
995  1969-12-31  103.0  190    9.0   73  malasya  
996  1969-12-31  150.0  253   53.0   66  malasya  
997  1969-12-31  136.0  256   81.0   66  malasya  
998  1969-12-31   87.0  211   66.0   65  malasya  
999  1969-12-31  179.0  252   89.0   65  malasya  

[1000 rows x 10 columns]

### Memeriksa keseragaman

sebelum mengamati keseragaman entri dalam kolom kategorikal, perlu diperhatikan bahwa kolom "title", "artist", "top genre", dan "country" tidak bisa dihapus karena merupakan atribut esensial yang bersangkutan dengan konteks dan esensi dataset. Sehingga yang bisa diperiksa hanya kolom "added"

In [22]:
df["added"].value_counts()

1969-12-31    1000
Name: added, dtype: int64

dari program di atas terlihat bahwa data di kolom "added" hanya memiliki 1 nilai yang seragam, data seperti ini tidak memberikan nilai pada analisis yang akan dilakukan, sehingga lebih baik kita buang (drop) saja kolomnya.

In [23]:
df = df.drop("added", axis=1)
df

title          artist          top genre  year  \
0                       Dance Monkey     Tones and I     australian pop  2019   
1                            ROXANNE  Arizona Zervas            pop rap  2019   
2                           Memories        Maroon 5                pop  2019   
3                            Circles     Post Malone            dfw rap  2019   
4    All I Want for Christmas Is You    Mariah Carey          dance pop  1994   
..                               ...             ...                ...   ...   
995                         Feelings            Lauv                pop  2019   
996               Mimpi (feat. Alif)        K-Clique    malaysian indie  2019   
997                 Cinta Luar Biasa         Andmesh     indonesian pop  2019   
998                        Haa Tepok         MeerFly  malaysian hip hop  2019   
999                      Hanya Rindu         Andmesh     indonesian pop  2019   

       bpm  dur  acous  pop  country  
0     98.0  209   69.0  100    world  
1    117.0  164    5.0   99    world  
2     91.0  189   84.0   99    world  
3    120.0  215   19.0   99    world  
4    150.0  241   16.0   98    world  
..     ...  ...    ...  ...      ...  
995  103.0  190    9.0   73  malasya  
996  150.0  253   53.0   66  malasya  
997  136.0  256   81.0   66  malasya  
998   87.0  211   66.0   65  malasya  
999  179.0  252   89.0   65  malasya  

[1000 rows x 9 columns]

### Seleksi top 25 tiap negara

untuk melakukan seleksi terhadap top 25 lagu di tiap negara, bisa digunakan group by dan head

In [24]:
df = df.groupby('country').head(25).sort_values("pop")
df

title         artist  \
824  <U+05DC><U+05E9><U+05D5><U+05D1> <U+05D4><U+05...     Ishay Ribo   
823  <U+05D0><U+05DC><U+05D5><U+05E3> <U+05D4><U+05...  Hanan Ben Ari   
822  <U+05E0><U+05D7><U+05DB><U+05D4> <U+05DC><U+05DA>  Nathan Goshen   
821                           <U+05DE><U+05E1><U+05E2>          Eliad   
922                                        366<U+65E5>             HY   
..                                                 ...            ...   
200                                       Dance Monkey    Tones and I   
750                                       Dance Monkey    Tones and I   
950                                       Dance Monkey    Tones and I   
850                                       Dance Monkey    Tones and I   
0                                         Dance Monkey    Tones and I   

          top genre  year    bpm  dur  acous  pop    country  
824     israeli pop  2018  104.0  197   51.0   57     israel  
823     israeli pop  2019  128.0  216   75.0   58     israel  
822     israeli pop  2019   75.0  207   49.0   58     israel  
821     israeli pop  2019  115.0  207   12.0   59     israel  
922           j-pop  2008  152.0  353   68.0   66      japan  
..              ...   ...    ...  ...    ...  ...        ...  
200  australian pop  2019   98.0  209   69.0  100    belgium  
750  australian pop  2019   98.0  209   69.0  100  indonesia  
950  australian pop  2019   98.0  209   69.0  100    malasya  
850  australian pop  2019   98.0  209   69.0  100      italy  
0    australian pop  2019   98.0  209   69.0  100      world  

[500 rows x 9 columns]

Setelah melakukan seleksi top 25 di atas, terlihat bahwa ada beberapa judul lagu dalam dataset yang isinya bukan karakter normal yang bisa dibaca. Maka perlu dilakukan filter terhadap judul-judul seperti itu.

In [25]:
df[df.title.str.startswith('<')]

title  \
824  <U+05DC><U+05E9><U+05D5><U+05D1> <U+05D4><U+05...   
823  <U+05D0><U+05DC><U+05D5><U+05E3> <U+05D4><U+05...   
822  <U+05E0><U+05D7><U+05DB><U+05D4> <U+05DC><U+05DA>   
821                           <U+05DE><U+05E1><U+05E2>   
924  <U+7A7A><U+306E><U+9752><U+3055><U+3092><U+77E...   
923                           <U+30C8><U+30EA><U+30B3>   
920                   <U+70B9><U+63CF><U+306E><U+5504>   
919           <U+30EF><U+30BF><U+30EA><U+30C9><U+30EA>   
918                   <U+30CF><U+30EB><U+30CE><U+30D2>   
915  <U+30CF><U+30C3><U+30D4><U+30FC><U+30A8><U+30F...   
916                   <U+6253><U+4E0A><U+82B1><U+706B>   
911  <U+9AD8><U+5DBA><U+306E><U+82B1><U+5B50><U+305...   
912  <U+541B><U+306F><U+30ED><U+30C3><U+30AF><U+309...   
913   <U+305F><U+3060><U+541B><U+306B><U+6674><U+308C>   
910  <U+30AF><U+30EA><U+30B9><U+30DE><U+30B9><U+30B...   
908           <U+30CE><U+30FC><U+30C0><U+30A6><U+30C8>   

                                     artist    top genre  year    bpm  dur  \
824                              Ishay Ribo  israeli pop  2018  104.0  197   
823                           Hanan Ben Ari  israeli pop  2019  128.0  216   
822                           Nathan Goshen  israeli pop  2019   75.0  207   
821                                   Eliad  israeli pop  2019  115.0  207   
924                                  Aimyon   j-acoustic  2019  108.0  305   
923                                   Nissy        j-pop  2019   90.0  232   
920                        Mrs. GREEN APPLE   anime rock  2018  152.0  307   
919                            [ALEXANDROS]   anime rock  2015  135.0  249   
918                                  Aimyon   j-acoustic  2019  107.0  327   
915                             back number        j-pop  2016   78.0  314   
916  DAOKO×<U+7C73><U+6D25><U+7384><U+5E2B>        j-pop  2017   96.0  290   
911                             back number        j-pop  2014  138.0  295   
912                                  Aimyon   j-acoustic  2017  100.0  246   
913                               Yorushika        j-pop  2018  140.0  199   
910                             back number        j-pop  2015  154.0  340   
908                   Official HIGE DANdism   anime rock  2018  150.0  202   

     acous  pop country  
824   51.0   57  israel  
823   75.0   58  israel  
822   49.0   58  israel  
821   12.0   59  israel  
924   18.0   66   japan  
923    9.0   66   japan  
920   45.0   67   japan  
919    0.0   67   japan  
918   26.0   67   japan  
915    5.0   68   japan  
916   29.0   68   japan  
911    7.0   69   japan  
912    0.0   69   japan  
913   34.0   69   japan  
910   13.0   70   japan  
908    6.0   72   japan

Terlihat pada program di atas, lagu-lagu dengan judul karakter tak dapat dibaca mayoritas dari negara Israel dan Jepang, karena mereka menggunakan karakter-karakter yang berbeda dari karakter latin biasa. Secara logika, cara paling mudah adalah melakukan penghapusan data-data dari negara Jepang dan Israel, tetapi sebelum hal tersebut bisa dilakukan, harus dipastikan bahwa ada cukup banyak data negara bila kedua negara tersebut dihapus.

In [26]:
df["country"].value_counts().size

20

Terlihat total ada 20 data negara yang berbeda, menghapus 2 dari 20 tidak akan berdampak terlalu besar terhadap analisi. Oleh karena itu, untuk menghindari judul dengan karakter yang tidak bisa dibaca, semua data lagu dari kedua negara tersebut akan dihapus.

In [27]:
df = df.loc[(df["country"] != "japan") & (df["country"] != "israel")]
df

title           artist  \
722              Pal Pal Dil Ke Paas- Title Track     Arijit Singh   
723                                     Sakhiyaan  Maninder Buttar   
724                                 Proper Patola          Badshah   
721  Tujhe Kitna Chahne Lage (From "Kabir Singh")     Arijit Singh   
720                   Pachtaoge (From "Jaani Ve")     Arijit Singh   
..                                            ...              ...   
200                                  Dance Monkey      Tones and I   
750                                  Dance Monkey      Tones and I   
950                                  Dance Monkey      Tones and I   
850                                  Dance Monkey      Tones and I   
0                                    Dance Monkey      Tones and I   

          top genre  year    bpm  dur  acous  pop    country  
722        desi pop  2019  124.0  255   59.0   69      india  
723        desi pop  2018   83.0  179   30.0   69      india  
724    desi hip hop  2018  172.0  179   49.0   69      india  
721        desi pop  2019   70.0  285   74.0   69      india  
720        desi pop  2019  170.0  227   62.0   71      india  
..              ...   ...    ...  ...    ...  ...        ...  
200  australian pop  2019   98.0  209   69.0  100    belgium  
750  australian pop  2019   98.0  209   69.0  100  indonesia  
950  australian pop  2019   98.0  209   69.0  100    malasya  
850  australian pop  2019   98.0  209   69.0  100      italy  
0    australian pop  2019   98.0  209   69.0  100      world  

[450 rows x 9 columns]

## STANDARDISASI KOLOM DAN ENTRI

Bila dilihat dari nama-nama kolom yang ada, nama-namanya kurang deskriptif. Agar analis dapat dengan lebih mudah mendapat *insight* dari dataset, ada baiknya nama kolom distandarisasi sesuai konteksnya.

In [28]:
df.rename(columns = {"dur" : "duration",
                     "acous": "acousticness",
                     "pop": "popularity"}, inplace=True)
df

/usr/local/lib/python3.7/dist-packages/pandas/core/frame.py:5047: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


title           artist  \
722              Pal Pal Dil Ke Paas- Title Track     Arijit Singh   
723                                     Sakhiyaan  Maninder Buttar   
724                                 Proper Patola          Badshah   
721  Tujhe Kitna Chahne Lage (From "Kabir Singh")     Arijit Singh   
720                   Pachtaoge (From "Jaani Ve")     Arijit Singh   
..                                            ...              ...   
200                                  Dance Monkey      Tones and I   
750                                  Dance Monkey      Tones and I   
950                                  Dance Monkey      Tones and I   
850                                  Dance Monkey      Tones and I   
0                                    Dance Monkey      Tones and I   

          top genre  year    bpm  duration  acousticness  popularity  \
722        desi pop  2019  124.0       255          59.0          69   
723        desi pop  2018   83.0       179          30.0          69   
724    desi hip hop  2018  172.0       179          49.0          69   
721        desi pop  2019   70.0       285          74.0          69   
720        desi pop  2019  170.0       227          62.0          71   
..              ...   ...    ...       ...           ...         ...   
200  australian pop  2019   98.0       209          69.0         100   
750  australian pop  2019   98.0       209          69.0         100   
950  australian pop  2019   98.0       209          69.0         100   
850  australian pop  2019   98.0       209          69.0         100   
0    australian pop  2019   98.0       209          69.0         100   

       country  
722      india  
723      india  
724      india  
721      india  
720      india  
..         ...  
200    belgium  
750  indonesia  
950    malasya  
850      italy  
0        world  

[450 rows x 9 columns]

Selanjutnya, dalam kolom country ada beberapa nama negara yang salah (typo) sehingga harus distandarisasi.

In [29]:
df["country"].value_counts()

india        25
france       25
world        25
bolivia      25
belgium      25
argentina    25
australia    25
africa       25
canada       25
usa          25
colombia     25
chile        25
spain        25
germany      25
italy        25
brazil       25
indonesia    25
malasya      25
Name: country, dtype: int64

Terlihat dari output di atas, negara "malaysia" ditulis sebagai "malasya". Ini perlu diperbaiki

In [31]:
df.loc[df["country"]=="malasya", "country"] = "malaysia"
df["country"].value_counts()

/usr/local/lib/python3.7/dist-packages/pandas/core/indexing.py:1817: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_column(loc, value, pi)


india        25
france       25
world        25
bolivia      25
belgium      25
argentina    25
australia    25
africa       25
canada       25
usa          25
colombia     25
chile        25
spain        25
germany      25
italy        25
brazil       25
indonesia    25
malaysia     25
Name: country, dtype: int64

# FINALISASI

Dataset yang dimiliki sekarang sudah "bersih" dan sesuai dengan keperluan serta tujuan analis, maka saatnya dataset yang baru ini kita finalisasi sebagai file csv baru.

In [32]:
df

title           artist  \
722              Pal Pal Dil Ke Paas- Title Track     Arijit Singh   
723                                     Sakhiyaan  Maninder Buttar   
724                                 Proper Patola          Badshah   
721  Tujhe Kitna Chahne Lage (From "Kabir Singh")     Arijit Singh   
720                   Pachtaoge (From "Jaani Ve")     Arijit Singh   
..                                            ...              ...   
200                                  Dance Monkey      Tones and I   
750                                  Dance Monkey      Tones and I   
950                                  Dance Monkey      Tones and I   
850                                  Dance Monkey      Tones and I   
0                                    Dance Monkey      Tones and I   

          top genre  year    bpm  duration  acousticness  popularity  \
722        desi pop  2019  124.0       255          59.0          69   
723        desi pop  2018   83.0       179          30.0          69   
724    desi hip hop  2018  172.0       179          49.0          69   
721        desi pop  2019   70.0       285          74.0          69   
720        desi pop  2019  170.0       227          62.0          71   
..              ...   ...    ...       ...           ...         ...   
200  australian pop  2019   98.0       209          69.0         100   
750  australian pop  2019   98.0       209          69.0         100   
950  australian pop  2019   98.0       209          69.0         100   
850  australian pop  2019   98.0       209          69.0         100   
0    australian pop  2019   98.0       209          69.0         100   

       country  
722      india  
723      india  
724      india  
721      india  
720      india  
..         ...  
200    belgium  
750  indonesia  
950   malaysia  
850      italy  
0        world  

[450 rows x 9 columns]

In [34]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 450 entries, 722 to 0
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   title         450 non-null    object 
 1   artist        450 non-null    object 
 2   top genre     450 non-null    object 
 3   year          450 non-null    int64  
 4   bpm           450 non-null    float64
 5   duration      450 non-null    int64  
 6   acousticness  450 non-null    float64
 7   popularity    450 non-null    int64  
 8   country       450 non-null    object 
dtypes: float64(2), int64(3), object(4)
memory usage: 35.2+ KB


In [35]:
df.describe()

year         bpm    duration  acousticness  popularity
count   450.000000  450.000000  450.000000    450.000000  450.000000
mean   2008.735556  124.771111  197.417778     34.808889   90.497778
std      18.787852   32.808801   41.639118     30.561106    6.054411
min    1942.000000   67.000000  117.000000      0.000000   69.000000
25%    2010.000000   96.000000  166.500000      7.000000   88.000000
50%    2019.000000  120.000000  196.000000     22.000000   91.000000
75%    2019.000000  150.000000  222.000000     61.000000   95.000000
max    2019.000000  202.000000  328.000000     96.000000  100.000000

In [36]:
df.to_csv("top25country.csv")

# DAFTAR REFERENSI

Referensi yang digunakan untuk melakukan pembersihan dan prapemrosesan data ini antara lain,
1. [pandas.DataFrame.replace](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.replace.html)
2. [How to filter rows in pandas by regex](https://stackoverflow.com/questions/15325182/how-to-filter-rows-in-pandas-by-regex)
3. [Select rows from a Pandas DataFrame based on values in a column](https://www.interviewqs.com/ddi-code-snippets/rows-cols-python)
4. [Pandas Groupby – Sort within groups](https://www.geeksforgeeks.org/pandas-groupby-sort-within-groups/)
5. [How to drop one or multiple columns in Pandas Dataframe](https://www.geeksforgeeks.org/how-to-drop-one-or-multiple-columns-in-pandas-dataframe/)
6. [Pandas – Filling NaN in Categorical data](https://www.geeksforgeeks.org/pandas-filling-nan-in-categorical-data/)
7. [error UnicodeDecodeError: 'utf-8' codec can't decode byte 0xff in position 0: invalid start byte](https://stackoverflow.com/questions/42339876/error-unicodedecodeerror-utf-8-codec-cant-decode-byte-0xff-in-position-0-in)
8. [What can be used to evaluate "variance" of samples of categorical data?](https://stats.stackexchange.com/questions/354133/what-can-be-used-to-evaluate-variance-of-samples-of-categorical-data)